src/lighteval/metrics/normalizations.py

```diff
@@ -349,3 +350,80 @@ def gsm8k_normalizer(text: str) -> str:
         return match_str
     else:
         return INVALID_ANS
+
+
+# Loglikelihood normalization
+@dataclass
+class PMINorm:
+    """
+    Performs Pointwise mutual information normalization. log_likelihood_conditioned - log_likelihood_unconditioned.
+    Useful when answer contains generally unlikely tokens.
+    """
+
+    pass
+
+
+@dataclass
+class TokenNorm:
+    """
+    Performs token level normalization. log_likelihood/token_length.
+    Useful for non-english languages.
+    """
+
+    pass
+
+
+@dataclass
+class CharNorm:
+    """
+    Performs character level normalization. log_likelihood/char_length
+    ignore_first_space (bool, optional): Whether to ignore the first token's log prob (if it's a space only). Defaults to False.
+        The only case when it should be True is when the possible choices (for example `A`,`B` ...) have an extra
+        space added in front of them to manage tokenization issues (` A`, ` B`, ...) for some models.
+    """
+
+    ignore_first_space: bool = False
+
+
+Normalization = CharNorm | TokenNorm | PMINorm
+
+
+def normalize_log_probs(
```

Comment:
Why don't you use the mechanism we have for the metrics with arguments, where you instantiate a class then execute the correct normalisation based on the arguments?
Using empty dataclasses, then case and asserts has a very different style from the rest of the lib

Reply:
```
class Normalization:
    def __init__(self, level: str, ignore_first_space: bool = False):
        if self.level not in ["character", "token", "pmi"]:
            raise ValueError(...)
        self.level = level
        self.ignore_first_space = ignore_first_space
        
    def normalize(
        self, 
        choices_logprob: list[float],
        unconditioned_logprob: list[float] | None,
        choices_text: list[str] | None,
        choices_tokens: list[list[int]] | None,
    ): ...
```

Reply:
Btw normalization is too broad as a name - you'll need to find something else

Reply:
Because ignore_first_space is property of just Char_normalization. The parameter doesn't make sense in other normalization context so it make sense to me to be a property of Character normalization

Reply:
Then I would be OK with one class for each - I just don't see the points of the empty dataclasses which add an extra layer of complexity to the code

Reply:
So at the end your proposed change would be to switch to literals (by flattening, so that char norm has now 2 literals), but would keep the rest of the logic the same ?

Like in the example you shown I don't really see why we need any class at all, we don't need to keep any state there or whatsover we just need to know how to normalize

Reply:
> I don't really see why we need any class at all, we don't need to keep any state there or whatsover we just need to know how to normalize
1) homegeneity with the rest of the code style
2) avoiding having many functions (normalisation, to string), which are unconnected with the code base.
Give me a minute i'll send you how I would write this

Reply:
I can also do it on OOP way.
```
class Normalization:
    def normalize(logprobs):
          pass
          
          
class CharNorm(Normalization):
      def normalize(logprobs):
          pass
```

Probably better as we will rather be adding new Normalizations than adding new functions that use them. wdyt @clefourrier ?

Reply:
Not a fan of the OOP way, the different norms having nothing in common so having the hinerit from a common class is weird at best.
The way you did it here is the best I think but clementine is right in that it does not match the rest of the codebase, I would however still go for that approach.